LabExT/Movement/Stage3DSmarAct.py

```diff
@@ -0,0 +1,625 @@
+                    ))):
+                self._acceleration = umps2
+
+        @property
+        def movement_mode(self) -> MovementType:
+            """Returns movement mode of channel: Either RELATIVE or ABSOLUTE"""
+            return self._movement_mode
+
+        @movement_mode.setter
+        def movement_mode(self, mode: MovementType) -> None:
+            """Set movement mode of channel: Either RELATIVE or ABSOLUTE"""
+            if mode.value not in MovementType._value2member_map_:
+                raise ValueError("Invalid movement mode {}".format(str(mode)))
+            self._movement_mode = mode
+
+        # Movement
+
+        def move(
+                self,
+                diff: float,
+                mode=MovementType.RELATIVE,
+                wait_for_stopping=True) -> None:
+            """Moves the channel with the specified movement type by the value diff
+
+            Parameters
+            ----------
+            diff : float
+                Channel movement measured in micrometers.
+            mode : MovementType
+                Channel movement type
+            """
+            self.movement_mode = mode
+            if self.movement_mode == MovementType.RELATIVE:
+                self._move_relative(diff, wait_for_stopping)
+            elif self.movement_mode == MovementType.ABSOLUTE:
+                self._move_absolute(diff, wait_for_stopping)
+
+        def find_reference_mark(self):
+            """Moves the channel to a known physical position, by searching for the reference mark"""
+            self._stage._exit_if_error(
+                MCSC.SA_FindReferenceMark_S(
+                    self._stage.handle,
+                    self._handle,
+                    MCSC.SA_BACKWARD_DIRECTION,
+                    0,
+                    MCSC.SA_AUTO_ZERO
+                ))
+
+        # Helper functions
+
+        def _move_relative(self, diff: float, wait_for_stopping: bool):
+            self._stage._exit_if_error(
+                MCSC.SA_GotoPositionRelative_S(
+                    self._stage.handle,
+                    self._handle,
+                    ct.c_int(int(self._to_nanometer(diff))),
+                    0
+                ))
+            if wait_for_stopping:
+                self._wait_for_stopping()
+
+        def _move_absolute(self, pos: float, wait_for_stopping: bool):
+            self._stage._exit_if_error(
+                MCSC.SA_GotoPositionAbsolute_S(
+                    self._stage.handle,
+                    self._handle,
+                    ct.c_int(int(self._to_nanometer(pos))),
+                    0
+                ))
+            if wait_for_stopping:
+                self._wait_for_stopping()
+
+        def _humanize_status(self, status_code: int) -> str:
+            if(status_code in self.STATUS_CODES):
+                return self.STATUS_CODES[status_code]
+
+            return "Unknown status code: " + str(status_code)
+
+        def _wait_for_stopping(self):
```

Comment:
I suggest adding an optional timeout feature here. If the movement takes longer than a specified amount of seconds, you can raise a RuntimeError here.

Reply:
I agree. But what would be a reasonable time to wait? Assuming the stage is moving at very low speed over a long distance, it would take longer for the stage to come to a stop. We could ask the user after a certain time through a dialog if he wants to stop the stage.

Reply:
So my idea was that `Stage.move_relative` and `Stage.move_absolute` have a timeout feature that you can enable (bool!). When enabled, Python would throw a RuntimeError if the stage takes longer for the required movement than we would expect based on the theoretical calculation with acceleration and speed + 20% overhead.

But since we are still just refactoring, dont worry about it, its fine.